Automatically create shortlinks for sharing in the love

dencold opened the pull request:
For hackathon 23, we wanted to enable shortlinks that could be used to easily share love that someone has created. In order to do this currently, most people hack around in the browser's URL to add query parameters and copy & paste after they see that it's working. With this PR, the lovemachine will do this automatically for you.

For more details, see [HACK-1765](https://jira.yelpcorp.com/browse/HACK-1765) (must be on Yelp internal network).

This also will fix several long-standing issues on the Love jira board:
- [LOVE-80](https://jira.yelpcorp.com/browse/LOVE-80)
- [LOVE-24](https://jira.yelpcorp.com/projects/LOVE/issues/LOVE-24)

I'd love to get this PR merged in by Monday so we can send it out for people to use in sharing Hackathon love. Also, I'm happy to deploy this into the GAE account if someone can add me onto the maintainer/owner list.

For a visual of what the changes look like, see this screenshot:
![Love Sent](https://raw.githubusercontent.com/dencold/love/love-links/screenshots/LoveSent.png)

For reference, here is who worked on this PR, along with our Yelp employee handles:
- @ajlau (Yelp = alau)
- alinar (same Yelp handle)
- @dencold (Yelp = dmc)

Thanks!!!!

rockdog commented on views/web.py:
```diff
@@ -215,17 +280,21 @@ def love():
             flash(exc.user_message, 'error')
         else:
             flash(exc.user_message)
 
```
I think there is a redirect + message missing here.

dencold replied:
agreed. alinar, can you fix this?

transcedentalia replied:
Ack.

dencold replied:
This still isn't fixed...I will take care of it directly.

dencold commented on logic/love_link.py:
```diff
@@ -0,0 +1,64 @@
+    if (loveLink is None):
+        raise NoSuchLoveLink("Couldn't Love Link with id {}".format(hash_key))
+
+    employee = Employee.get_key_for_username(recipient)
```
@ajlau did you know how aliases fold into this? I thought you mentioned checking on this. Let me/@rockdog know.

dencold replied:
@ajlau do you know how aliases work with the love link logic? I think you mentioned that you checked on this during hackathon, but can't remember.

ajlau replied:
@dencold, we probably want to use `name_for_alias()` or write a new function to resolve alias/username to employee object. https://github.com/Yelp/love/blob/master/logic/alias.py#L25

dencold commented on views/web.py:
```diff
@@ -88,6 +92,34 @@ def me_or_explore(user):
+
 @app.route('/explore', methods=['GET'])
 @user_required
 def explore():
```
@rockdog yeah, for sure. alinar. would you mind pulling these? I think you had added them in?

transcedentalia replied:
Sure, sorry for that. :D

dencold commented on views/tasks.py:
```diff
@@ -65,3 +66,9 @@ def notify_subscribers():
     return Response(status=200)
+
+
+@app.route('/tasks/lovelinks/cleanup', methods=['GET'])
```
agree with this. alinar, I thought we talked about GET/POST for the endpoint. was there a specific reason you used a GET here?

transcedentalia replied:
I remember reading somewhere in the documentation that only GET is supported. Maybe I'm wrong. To be honest, GET doesn't make too much sense.

ajlau replied:
Documentation says GAE cron service only supports GET. https://cloud.google.com/appengine/docs/standard/python/config/cron

rockdog replied:
You are right. There is even a comment in tasks.py. :)

dencold commented:
Hi @rockdog thanks for the quick review! I'll try to answer each of your points, above.

> Can you explain a little more how this feature changes existing flows? What are the entry and exit points?

Right now, flow is Send Love -> POST /love -> Redirect to /sent. This new "sent" landing page displays the screenshot you see above. We figured it'd be a better experience for the user to go to a landing page instead of trying to stuff more information into the flash message and redirect back home.

>It seems one has to actually send love in order to create a love link? I could also imagine a second button on the existing form to generate a link without having to send love.

Yes, that's correct. We plan to add a button here, just ran out of time during the hackathon. We wanted to focus on the path that didn't require the user to take any active action, they get the links for free right now and can decide whether or not they'd like to use them. We'll add a button for this soon.

> Why is /l/{:love_link_id} a separate page? Now the love send form is duplicated it seems. Couldn‘t the /l/{:love_link_id} just redirect to the existing homepage, set recipients and message accordingly and that way pre-fill the form and re-use existing functionality?

That's a good observation. One thing that we want to work on down the road is having the ability to "edit" these short links. A good use case here is when someone sends out a love link, but accidentally left someone off. Then someone else hops on the email thread and is like "Hey, here is the corrected love link". Now there are multiple emails going out. It would be nice if someone could come to page, see that someone was missing, and then just add them directly without the need to send an update email. Everyone else who uses the short link from that point on will get the correct recipient list. No need to spam people with additional emails.

> I don‘t see how this change creates permalinks to individual loves. Could you elaborate a bit. I would expect something like /love/{:love_id}.

Not sure I understand? When you click on /l/<hash> you get a direct view of the love that was sent. btw these aren't really "permalinks" we clean up any links that are older than 30 days.

> For the /sent page you are displaying the user avatars differently compared to any other page/email. Wouldn‘t it make sense to either use the existing way of displaying user avatars or change it everywhere?

Yes. We wanted to just use y/facestyles here, but didn't want to link into something that was internal only to yelp, so I created my own css to build up the avatars. It sounds like if I just reuse the styles from facestyles it would be better? I can work on that.

> Please add tests for all this new functionality.

But it's a HACKathon!!!! Who needs tests when we're HACKING!!?!?!?!?! Hehehehehhe. Thanks for pointing this out. I will get the tests written for the new functionality. ;)

dencold commented on views/web.py:
```diff
@@ -286,6 +280,7 @@ def love():
             flash('Secret love sent to {}!'.format(real_display_str))
             return redirect(url_for('home'))
         else:
+            flash('Love sent to {}!'.format(real_display_str))
```
this shouldn't be done here. /sent doesn't consume flash messages (we know that if I user gets to the /sent endpoint, they have successfully sent a message and we tell them that directly on the template). Right now, this flash will get consumed the next time a user hits the homepage which is *super* confusing.

transcedentalia replied:
I think I've misunderstood @rockdog's comment above  :"I think there is a redirect + message missing here.". Removing the flash message.

dencold commented on views/web.py:
```diff
@@ -286,6 +280,7 @@ def love():
+            flash('Love sent to {}!'.format(real_display_str))
             hash_key = link_id if link_id else create_love_link(real_display_str, message).hash_key
             return redirect(url_for('sent', message=message, recipients=real_display_str, link_id=hash_key))
 
```
the exception below still isn't being handled with a proper redirect. You should have `return redirect(url_for('home'))` at the bottom of the caught exception. Otherwise the user gets a 500 and these flashes don't go anywhere.

ajlau commented on logic/love_link.py:
```diff
@@ -0,0 +1,64 @@
+
+    employee = Employee.get_key_for_username(recipient)
+    if employee is None:
+        raise NoSuchEmployee('Couldn\'t find this user')
```
@dencold We can remove this check since `Employee.get_key_for_username(recipient)` already does this and should never return `None`

dencold replied:
@ajlau nice! will remove that, thanks!

dencold commented:
@rockdog we made some updates today. When you get a chance please give it another look-over and drop us a line on where we should focus next.

rockdog commented on cron.yaml:
```diff
@@ -0,0 +1,5 @@
+cron:
```
Since we already have a cron.yaml.example we should either
- get the contents from that example file into this cron.yaml and remove the example. This would also require to update the *make deploy* target Makefile because right now using cron is optional.
or
- move this into the example, so that people can decide whether to remove old links or keep them

I would vote for moving this into the example file + include a comment what this is for.

dencold replied:
agree with moving to an example file with comments. @transcedentalia mind doing the needful here?

dencold commented:
Thanks so much for the thoughtful comments @rockdog. I'm working on pulling together the rest of the tests. Hope to have it done this AM. 

Do you think it's possible to get a shipit today and be rolled out to AppEngine? Again, I'm happy to do the deployment myself, just need credentials.

Thanks!

dencold commented:
Hi @rockdog I've added in all the tests. Let me know if this lgty. I know it's getting late in your TZ, so I understand if you don't want to do this tonight ;) Let me know when you have a chance to review.

rockdog commented on tests/views/web_test.py:
```diff
@@ -269,6 +274,61 @@ def test_index_with_recipient_and_message(self):
+        self.assertIsNotNone(response.context['current_time'])
+        self.assertEqual(response.context['current_user'], self.current_user)
+        self.assertIsNotNone(response.context['loved'])
+        self.assertEqual(response.context['url'], config.APP_BASE_URL + 'l/cn23sx')
```
This test is failing on TravisCI

`Traceback (most recent call last):
  File "/home/travis/build/Yelp/love/tests/views/web_test.py", line 300, in test_sent_with_args
    self.assertEqual(response.context['url'], config.APP_BASE_URL + 'l/cn23sx')
AssertionError: 'https://PROJECT_ID.appspot.com//l/cn23sx' != 'https://PROJECT_ID.appspot.com/l/cn23sx'`

dencold replied:
yep. I pushed a fix right after I saw the failure. this now uses a mock to make sure this sort of thing isn't a problem again (old test relied on user's config.py).

dencold commented:
Yes, I think I should be a quick change to use a dropshadow for the text and we can get it matching in both places. Bear with me a sec.

dencold commented:
Okay, I think I've addressed all comments now @rockdog. I also went in and updated the `config.yaml.example` in my last commit since that never was addressed. Let me know if you see anything else that would prevent the merge. I really REALLY appreciate all of the time you've given us on this. :heart: :heart: :heart:

dencold commented:
@rockdog++ you are awesome. THANK YOU!!!
